### ISSUE TITLE
New MLP attribution

### ISSUE THREAD

**Issue by MonZop** [MonZop/BioBlender#22]

**Comment by MonZop** [comment]
Also consider that there is a quite short space of min-max distance between the vertex point and the closest atom

**Comment by zeffii** [comment]
Maybe use a proxy mesh, all vertices of which represent the cloud of atoms from the .pdb, the indices of the vertices will reflect the kind of atom. For example:
```
# keys would be generated once at runtime
index_mapper = {
    0: "O",     # 0 <= idx <= 200
    201: "H",   # 201 <= idx <= 320
    321: "N"    # 321 <= idx <= last index
}

```
In a K-dimensional Tree (KDTree) we add a list of vectors and indices. For every point on the MLP Surface mesh you can make a call to KDTree and get the index of the closest vertex, we then know the atom type.

This might not be the fastest solution, but it is the easiest to code and therefore becomes a benchmark to speed test subsequent iterations.

@MonZop 
> Also consider that there is a quite short space of min-max distance between the vertex point and the closest atom

OK, before starting to write code can you elaborate on which conditions this imposes?

**Comment by MonZop** [comment]
@zeffii 
>Maybe use a proxy mesh, all vertices of which represent the cloud of atoms from the .pdb, the indices of the vertices will reflect the kind of atom. 

The lipophilic value is not directly correspondent to the atom: it depends on how it is connected to other atoms. There is a library that specify the lipophilic value for every atom of all aminoacids and nucleic acids: this is the reference libray 
>For every point on the MLP Surface mesh ...

I think you mean the surface mesh: the MLP is not calculated yet (or did I not understand what you mean?)

**Comment by MonZop** [comment]
@zeffii 
>OK, before starting to write code can you elaborate on which conditions this imposes?

The surface mesh is defined as 'Solvent Excluded Surface', also called 'Connolly surface', defined as the points of the surface of a spherical probe rolling on the VanderWaals atoms.
![molecularsurface](https://cloud.githubusercontent.com/assets/10231334/6349934/d1876bc8-bc2f-11e4-8979-9e5fb6d6c692.png)
Due to this definition, the min distance between a point of the mesh and an atom of the molecule is = the radius of the smallest atom; the max distance will be >= (radius of the largest atom+radius of the probe). All this might be irrelevant if the KDtree works in a way that does not imply exploring the neighborhood to look for the closest point of the 'atom cloud mesh'

**Comment by zeffii** [comment]
right, this does introduce a slight detour. will digest..

Some throwaway code:  https://gist.github.com/zeffii/2ddf35134d9f1026ead2  (just parking it here for reference)

**Comment by zeffii** [comment]
>  Due to this definition, the min distance between a point of the mesh and an atom of the molecule is = the radius of the smallest atom; the max distance will be >= (radius of the largest atom+radius of the probe). 

I see! do I understand this? :

Even when the nearest atom is found, we will need to know of that atom its position in the _amino acid_ / _nucleic acid_ of which it is part, then do the library lookup for the lipophilic value?

**Comment by MonZop** [comment]
Yes, this info is in the columns 14-15 (atom) and 18-20 (aminoacid) of the PDB file. Then there is the fi values library

**Comment by zeffii** [comment]
Will implement a static version first, then if that's acceptable consider the dynamic / animatable approach. Maybe later today, this evening.

**Comment by zeffii** [comment]
cool, every atom is asigned info on `BBInfo` attribute, f.ex
```BBInfo : ATOM    232  CG  ARG A  16       5.046  -1.171  -2.378  1.00  0.00           C```

**Comment by MonZop** [comment]
If this BBinfo can be used, it might be easy to substitute the Temp factor (last 0.00, column 61-66) with fi value, and then its easy to retrieve.

**Comment by zeffii** [comment]
if BBinfo is never overwritten that might be a nice solution indeed, and easy.

**Comment by zeffii** [comment]
some debug images :) In 06_1L2Y_4GE.pdb

![image](https://cloud.githubusercontent.com/assets/619340/6367546/b3bd7ede-bcd4-11e4-9963-f970f81578cb.png)

**Comment by zeffii** [comment]
more interim scripting: https://gist.github.com/zeffii/c864b5096bdc4a479703

**Comment by MonZop** [comment]
If I understand the number on each atom is the fi value associated, right?
If this works, I will test it on a few molecules on which the MLP is especially important, so we can evaluate the result

**Comment by zeffii** [comment]
>  If I understand the number on each atom is the fi value associated, right?

Yep! from the _values_fi_ lookup table, they are not yet remapped / normalized `[-3, +1] ->  [0,1]`.  So far this is like administrative work..

Next step is to implement the KDTree search and design the algorithm to test element locations and determine the nearest element with respect to the elements' vdwaals radius. It's easy to limit the search area by passing a maximum search radius (probably maximum vdwaals radius in the pdb)

If this explanation is a bit confusing, hopefully the code will be more clear.

**Comment by zeffii** [comment]
the red vertices indicate the which vertices the search is limited to by passing a radius
![radus](https://cloud.githubusercontent.com/assets/619340/6370630/cadbf866-bcf2-11e4-8f93-67612bfa19ff.gif)

**Comment by MonZop** [comment]
@zeffii 
>passing a maximum search radius (probably maximum vdwaals radius in the pdb)

maximum search radius is = max VdW radius + probe radius

**Comment by zeffii** [comment]
Right, i imagine there are vertices on the Connolly surface mesh which due to the blobby-ness of the surface are interpolation values and the nearest element is further away than the max VdW radius. 

Do you mean the _probe_radius_ as a buffer zone -- an extension of the search radius? If so, that makes sense to me. 

![image](https://cloud.githubusercontent.com/assets/619340/6371802/5a1ce1a0-bcff-11e4-9be6-01a08ad54f1b.png)

here P One is a point on the Connolly surface which has no nearest neighbour if the search radius isn't extended

**Comment by zeffii** [comment]
probe radius could be this progressive thing, which first searches 2 times the max VdW radius , in most cases this will probably find something , if not then I would find the parameters used for the surface blending radius , and use that + some fudge factor for numeric imprecision.

**Comment by zeffii** [comment]
or perhaps
```python
results = search_kdtree(mode=distance, distance=max_distance)
if not results:
   results = search_kdtree(mode=nearest)
```

**Comment by zeffii** [comment]
Either way. it will be interesting to see the results, seeing often points to pitfalls in an algorithm.. 

No ETA on this, i'll think about it and perhaps write code tonight.

**Comment by MonZop** [comment]
Just a sec. 
I am making a figure that will explain better

**Comment by MonZop** [comment]
![ses1](https://cloud.githubusercontent.com/assets/10231334/6372637/ecf21df0-bd05-11e4-80d3-47f76b4037f0.png)

Aplogies for the low qaulity of the drawing.
Please note that the molecular surface defined is not by the center of the probe, but by its surface.
The radius of the probe is typically 0.7 A, considered the radius of a water molecule (yes, we know that water is not spherical, but it is one of the approximations that everyone accepts). This radius is comparable to the VdW radius of most atoms, so the farthest possible point between is:
r = probe radius (tipycally = or just smaller than d)
d = atom radius
Let's approximate r=d (staying a little large)
M = d sqrt(3)

**Comment by zeffii** [comment]
https://gist.github.com/zeffii/95e09d1e94feb2c23a8a mostly just the _choice + compare_ code now (interim code)

**Comment by zeffii** [comment]
![image](https://cloud.githubusercontent.com/assets/619340/6390131/fb9c6eba-bda9-11e4-8423-f083c4d47731.png)
https://gist.github.com/zeffii/37de8fde581c4fd98243
I just noticed not all elements are in the `scale_vdw` dictionary. for testing purposes the purple represents failed lookups for the radii of elements.
```
radii not found: {'OE1', 'ND2', 'CZ3', 'OXT', 'CD', 'CZ2', 'CD2', 'OD1', 'NE1', 'OH', 'NH2', 'NH1', 'OD2', 'CE2'
, 'CZ', 'CG', 'CB', 'NE', 'CG1', 'CD1', 'CG2', 'CE', 'NZ', 'NE2', 'CE1', 'CH2', 'OG'}
```
the good patches suggest the basic algorithm does work :)

**Comment by zeffii** [comment]
please forgive in advance this, which is probably butchering chemistry terms..

These entities (NE, SD OE1...) need to be remapped back from specific _element in the chain_ to the _generic element_ , and take the vdw from the generic element?

**Comment by zeffii** [comment]
Reading a pdb, even big ones, is a relatively fast operation. Generating a remap from 
```
['OE1'] => ['O']
```
could be done for every pdb imported, i suspect a dictionary lookup will be sufficiently fast. If it proves to be slow alternatives aren't that much more effort.

```python
def get_specific_dict(filepath):
    s_dict = {}

    with open(filepath) as pdb_file:
        for line in pdb_file:
            if line.startswith('ATOM'):
                specific = line[12:16].strip()
                _element = line[76:78].strip()
                if not (specific in s_dict):
                    s_dict[specific] = _element

    return s_dict

fp = r'C:\Users\dealga\Documents\GitHub\BioBlender\Test_molecules\06_1L2Y_4GE.pdb'
sd = get_specific_dict(filepath=fp)
print(sd)
```
gives
```
{
  'OD2': 'O', 
  'CD1': 'C', 
  'CH2': 'C', 
  'CG': 'C', 
  'HH': 'H', 
  'HD13': 'H', 
  'CE': 'C',
  'CZ': 'C',
  'HZ2': 'H',
  'HB': 'H', 
  # ....etc
}
```

**Comment by zeffii** [comment]
![image](https://cloud.githubusercontent.com/assets/619340/6391918/a6beb4e0-bdbb-11e4-8a63-9fa83e9d6eb4.png)

**Comment by zeffii** [comment]
https://gist.github.com/zeffii/fa8f7530d28dcaca4c7a functional..  now to test on larger files..

**Comment by zeffii** [comment]
still some missing lookup.. slightly puzzling...
![bowszer](https://cloud.githubusercontent.com/assets/619340/6392004/6b6d9edc-bdbc-11e4-86c9-f505df947f5c.gif)

**Comment by zeffii** [comment]
heh it's nice to have some time to dedicate to this..
it turns out
```
nothing within 2.7 of indexed atom 1287
nothing within 2.7 of indexed atom 2681
nothing within 2.7 of indexed atom 4180
```
Where indexed atom means a vertex on the Connolly surface. and 2.7 is the rescaled world distance. In some message earlier in the thread I said this might happen, and that the next search for proximity could be a simple 'find the closest' or 'perform an extended search'.

Do you have any preference here?

**Comment by zeffii** [comment]
![jbg](https://cloud.githubusercontent.com/assets/619340/6393485/b1890510-bdca-11e4-86f2-73a6a3c731bd.gif)

**Comment by zeffii** [comment]
will add as separate python file in BioBlender folder, and import it and add the functionality to the Surface Panel, this will make testing easier. Think of a name  for the button.

**Comment by zeffii** [comment]
it can be tested in the MLP alternative branch
![bluefork](https://cloud.githubusercontent.com/assets/619340/6398694/97770686-bded-11e4-955d-2eafd0fcde5e.png)

**Comment by MonZop** [comment]
Hi, sorry for not follwing yesterday. I was in Bologna at ameeting , with very poor connection, barely enough to check the mail occasionally.
I see there is quite some progress, I will need some time to think about it, however some indications can help immediately. 

>I just noticed not all elements are in the scale_vdw dictionary. for testing purposes the purple represents failed lookups for the radii of elements.
06_1L2Y_4GE.pdb
radii not found: {'OE1', 'ND2', 'CZ3', 'OXT', 'CD', 'CZ2', 'CD2', 'OD1', 

>These entities (NE, SD OE1...) need to be remapped back from specific element in the chain to the generic element , and take the vdw from the generic element?

You are right: the atomic elemnt determines the VdW radius, and is in column 11 of the PDB (most biological elemnts are simple ones, defined by a single letter, with important exception), sometimes it is also reported in the last columns of the PBD ATOM line, where the exceptions become visible (Ca, Fe, Mg, etc)

**Comment by zeffii** [comment]
> sometimes it is also reported in the last columns of the PBD ATOM 

Right now the code assumes the presence of that last column, it was rather convenient - but you say this column is not guaranteed. then an alternative scheme is necessary.
![toasty](https://cloud.githubusercontent.com/assets/619340/6409677/3540db0c-be67-11e4-801e-03f302171fe6.PNG)

**Comment by MonZop** [comment]
@zeffii 
> will add as separate python file in BioBlender folder, and import it and add the functionality to the Surface Panel, this will make testing easier. Think of a name for the button.

What about 'Direct MLP mapping'
(meaning that it is not medaited by integration),
I need some time to consider if this is OK, it looks like integration is somehow necessary.
(i have a deadline today, and cannot spend musch time on this, unfortunatley)

**Comment by zeffii** [comment]
@MonZop  No pressure from me. Button names are easy to change but deserve careful consideration. 
A fast-integration mode may be possible, if you can point at a source for the math involved. I can imagine routines but only from an aesthetic point of view :)

**Comment by zeffii** [comment]
btw, i remove all lights from these renderings (even viewport views) to get a better idea of the flat colours of the vertex_col map. No lights means no shadows.

**Comment by zeffii** [comment]
open-access is wonderful :) http://www.biomedcentral.com/1471-2105/10/32

**Comment by MonZop** [comment]
Yes, indeed!
We also published BioBlender as Open Access in the same journal. 
http://www.biomedcentral.com/1471-2105/13/S4/S16
